### PR TITLE
Add new features

### PR DESCRIPTION
**Pull request by binarymaster**
- Add ability to detect routers with "empty string" pins
- Argument for output file
- Full-length pin brute-force (for non-standard pins like 222222)

### CODE REVIEW

**Comment by kimocoder** [comment]
Great work again, binarymaster! Thanks alot for this one :)

**Comment by binarymaster** [comment]
Added more useful features :)

**Comment by kimocoder** [comment]
@wiire-a merge this and bump version? thanks :)

**Comment by wiire-a** [comment]
@kimocoder will do tomorrow hopefully. I wanted to take a proper look at the changes before merging.

File: src/pixiewps.c [hunk 1]
```diff
@@ -919,6 +949,107 @@ uint_fast8_t crack(struct global *g, unsigned int *pin) {
...
+				memcpy(buffer + WPS_SECRET_NONCE_LEN + WPS_PSK_LEN + WPS_PKEY_LEN, wps->pkr, WPS_PKEY_LEN);
+				hmac_sha256(wps->authkey, WPS_AUTHKEY_LEN, buffer,
+					WPS_SECRET_NONCE_LEN + WPS_PSK_LEN + WPS_PKEY_LEN * 2, result);
+
+				if (memcmp(result, wps->e_hash1, WPS_HASH_LEN)) {
+					first_half++;
+				} else {
+					if (i == 0)
+					{
+						snprintf(pin, WPS_PIN_LEN + 1, "");
```
Comment (wiire-a) [inline src/pixiewps.c]: This should be (to avoid compiler warnings):
```C
pin[0] = '\0';
```

File: src/pixiewps.c [hunk 2]
```diff
@@ -919,6 +949,107 @@ uint_fast8_t crack(struct global *g, unsigned int *pin) {
...
+		memcpy(buffer + WPS_SECRET_NONCE_LEN + WPS_PSK_LEN + WPS_PKEY_LEN, wps->pkr, WPS_PKEY_LEN);
+		hmac_sha256(wps->authkey, WPS_AUTHKEY_LEN, buffer,
+			WPS_SECRET_NONCE_LEN + WPS_PSK_LEN + WPS_PKEY_LEN * 2, result);
+
+		if (!memcmp(result, wps->e_hash2, WPS_HASH_LEN)) {
+			/* Empty pin detected */
+			free(buffer);
+			free(result);
+
+			snprintf(pin, WPS_PIN_LEN + 1, "");
```
Comment (wiire-a) [inline src/pixiewps.c]: Same as above (to avoid compiler warnings):
```C
pin[0] = '\0';
```

File: src/pixiewps.c [hunk 3]
```diff
@@ -462,10 +474,12 @@ int main(int argc, char **argv) {
 
 	uint_fast8_t k = 0;
 	uint_fast8_t found_p_mode = NONE;
-	unsigned int pin;
+	char *pin;
```
Comment (wiire-a) [inline src/pixiewps.c]: Being just 8 + 1 characters in length I would avoid using dynamic memory allocation:
```C
char pin[WPS_PIN_LEN + 1];
```

File: src/pixiewps.c [hunk 4]
```diff
@@ -462,10 +474,12 @@ int main(int argc, char **argv) {
...
 	uint32_t seed;
 	uint32_t print_seed = 0;
 
+	pin = calloc(WPS_PIN_LEN + 1, 1); if (!pin) goto memory_err;
```
Comment (wiire-a) [inline src/pixiewps.c]: With the change proposed above this line would go.

In any case, to be more precise the line should be (although in C99 I believe sizeof(char) is always 1):
```C
pin = calloc(WPS_PIN_LEN + 1, sizeof(char)); if (!pin) goto memory_err;
```

File: src/pixiewps.c [hunk 5]
```diff
@@ -841,6 +859,7 @@ int main(int argc, char **argv) {
 		free(wps->warning);
 	}
 
+	free(pin);
```
Comment (wiire-a) [inline src/pixiewps.c]: With stack allocation this line would go (see above).

**Comment by binarymaster** [comment]
Thanks for the review! I fixed warnings and removed dynamic memory allocation for pin.

**Comment by wiire-a** [comment]
The casts (when calling `crack(..., ....)`):
```C
(char *)&pin
```
are pointeless, pin is already of type `char *`. Just:
```C
pin
```
is enough.

In C the name of the array is already a pointer to the first element of the array (in fact `pin` it's equal to `&pin[0]`)

**Comment by binarymaster** [comment]
Fixed :)

**Comment by wiire-a** [comment]
Thank you!

**Comment by kimocoder** [comment]
could you bump pixiewps version so it would be quickly included in Kali? thanks both of you

**Comment by wiire-a** [comment]
Done (v1.3). Hopefully they'll upgrade.

**Comment by kimocoder** [comment]
They will ;)

**Comment by kimocoder** [comment]
[here](https://bugs.kali.org/view.php?id=4293) it's assigned to Sophie. again, thanks both @wiire-a & @binarymaster great work.

**Pull request merged**